File: apps/chat_client.cpp
// Minimal chat client built entirely on the minitcp_* socket API.
// Reads lines from stdin, sends each to chat_server, prints the
// echoed reply. Runs on its own TUN device (default tun1) so it can
// genuinely connect to chat_server (on tun0) through the kernel's own
// IP forwarding between the two — a real two-party conversation over
// our own TCP implementation, not just against external tools.
#include "socket_api.h"

#include <arpa/inet.h>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <string>

namespace {

uint32_t parse_addr(const char* s) {
    struct in_addr a;
    if (inet_pton(AF_INET, s, &a) != 1) {
        std::fprintf(stderr, "invalid address: %s\n", s);
        std::exit(1);
    }
    return ntohl(a.s_addr);
}

}  // namespace

int main(int argc, char** argv) {
    uint16_t port = 8080;
    std::string tun_dev = "tun1";
    std::string self_addr_str = "10.0.1.2";
    std::string server_addr_str = "10.0.0.2";
    bool trace = false;

    for (int i = 1; i < argc; ++i) {
        if (std::strcmp(argv[i], "--port") == 0 && i + 1 < argc) {
            port = static_cast<uint16_t>(std::atoi(argv[++i]));
        } else if (std::strcmp(argv[i], "--tun") == 0 && i + 1 < argc) {
            tun_dev = argv[++i];
        } else if (std::strcmp(argv[i], "--addr") == 0 && i + 1 < argc) {
            self_addr_str = argv[++i];
        } else if (std::strcmp(argv[i], "--server") == 0 && i + 1 < argc) {
            server_addr_str = argv[++i];
        } else if (std::strcmp(argv[i], "--trace") == 0) {
            trace = true;
        }
    }

    uint32_t self_addr = parse_addr(self_addr_str.c_str());
    uint32_t server_addr = parse_addr(server_addr_str.c_str());

    if (!minitcp::minitcp_init(tun_dev, self_addr, trace)) {
        std::perror("minitcp_init");
        return 1;
    }

    minitcp::Socket* sock = minitcp::minitcp_socket();
    if (!minitcp::minitcp_connect(sock, server_addr, port)) {
        std::fprintf(stderr, "chat_client: failed to connect to %s:%u\n",
                     server_addr_str.c_str(), port);
        return 1;
    }
    std::printf("chat_client: connected to %s:%u -- type messages, Ctrl+D to quit\n",
               server_addr_str.c_str(), port);

    char line[1024];
    while (std::fgets(line, sizeof(line), stdin)) {
        size_t len = std::strlen(line);
        minitcp::minitcp_send(sock, line, len);

        char buf[2048];
        ssize_t n = minitcp::minitcp_recv(sock, buf, sizeof(buf));
        if (n <= 0) {
            std::printf("chat_client: server closed the connection\n");
            break;
        }
        std::fwrite(buf, 1, static_cast<size_t>(n), stdout);
        std::fflush(stdout);
    }

    minitcp::minitcp_close(sock);
    return 0;
}

File: src/bin/chat_client.rs
//! Minimal chat client built entirely on the `Stack` socket API. Reads
//! lines from stdin, sends each to chat_server, prints the echoed
//! reply. Runs on its own TUN device (default tun1) so it can
//! genuinely connect to chat_server (on tun0) through the kernel's own
//! IP forwarding between the two — a real two-party conversation over
//! our own TCP implementation, not just against external tools.
use std::io::{self, BufRead, Write};
use std::net::Ipv4Addr;

fn parse_addr(s: &str) -> u32 {
    match s.parse::<Ipv4Addr>() {
        Ok(a) => u32::from(a),
        Err(_) => {
            eprintln!("invalid address: {s}");
            std::process::exit(1);
        }
    }
}

fn main() {
    let mut port: u16 = 8080;
    let mut tun_dev = "tun1".to_string();
    let mut self_addr_str = "10.0.1.2".to_string();
    let mut server_addr_str = "10.0.0.2".to_string();
    let mut trace = false;

    let args: Vec<String> = std::env::args().collect();
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--port" if i + 1 < args.len() => {
                i += 1;
                port = args[i].parse().unwrap_or(8080);
            }
            "--tun" if i + 1 < args.len() => {
                i += 1;
                tun_dev = args[i].clone();
            }
            "--addr" if i + 1 < args.len() => {
                i += 1;
                self_addr_str = args[i].clone();
            }
            "--server" if i + 1 < args.len() => {
                i += 1;
                server_addr_str = args[i].clone();
            }
            "--trace" => trace = true,
            _ => {}
        }
        i += 1;
    }

    let self_addr = parse_addr(&self_addr_str);
    let server_addr = parse_addr(&server_addr_str);

    let mut stack = match minitcp::Stack::init(&tun_dev, self_addr, trace) {
        Some(s) => s,
        None => {
            eprintln!("minitcp_init: {}", io::Error::last_os_error());
            std::process::exit(1);
        }
    };

    let sock = stack.socket();
    if !stack.connect(sock, server_addr, port) {
        eprintln!("chat_client: failed to connect to {server_addr_str}:{port}");
        std::process::exit(1);
    }
    println!("chat_client: connected to {server_addr_str}:{port} -- type messages, Ctrl+D to quit");

    let stdin = io::stdin();
    for line in stdin.lock().lines() {
        let Ok(mut line) = line else { break };
        line.push('\n');
        stack.send(sock, line.as_bytes());

        let mut buf = [0u8; 2048];
        let n = stack.recv(sock, &mut buf);
        if n <= 0 {
            println!("chat_client: server closed the connection");
            break;
        }
        io::stdout().write_all(&buf[..n as usize]).ok();
        io::stdout().flush().ok();
    }

    stack.close(sock);
}
